resources/session02/http_server.py

```diff
@@ -1,12 +1,73 @@
 import socket
 import sys
+import mimetypes
+import os.path
+
+
+def response_ok(body=b"this is a pretty minimal response", mimetype=b"text/plain"):
+    """returns a basic HTTP response"""
+    mimetype = b"content-type:" + mimetype
+    resp = []
+    resp.append(b"HTTP/1.1 200 OK")
+    resp.append(mimetype)
+    resp.append(b"")
+    resp.append(body)
+    return b"\r\n".join(resp)
+
+
+def response_method_not_allowed():
+    """returns a 405 Method Not Allowed response"""
+    resp = []
+    resp.append("HTTP/1.1 405 Method Not Allowed")
+    resp.append("")
+    return "\r\n".join(resp).encode('utf8')
+
+
+def response_not_found():
+    """returns a 404 Not Found response"""
+    return b"HTTP/1.1 404 Not Found"
+
+
+def parse_request(request):
+    first_line = request.split("\r\n", 1)[0]
+    try:
+        method, uri, protocol = first_line.split()
+    # Chrome seems to be sending a zero-length request that breaks the split.
+    except ValueError:
+        print('The client tried to break me.  Request was: \'{}\''.format(request))
+        method = ''
+    if method != "GET":
+        raise NotImplementedError("We only accept GET")
+    return uri
+
+
+def resolve_uri(uri):
+    """Return appropriate content and a mime type"""
+    # Return the contents of a file if it exists
+    uri = os.getcwd() + uri
+    if os.path.isfile(uri):
+        mime_type, _ = mimetypes.guess_type(uri)
+        mime_type = mime_type.encode('utf8')
+        with open(uri, "rb") as content_file:
+            f = content_file.read()
+            content = bytearray(f)
+    # Return an index if a directory is requested
+    elif os.path.isdir(uri):
+        mime_type = b'text/plain'
+        filelist = os.listdir(uri)
+        filelist.sort
+        content = '\r\n'.join(filelist)
+        content = content.encode('utf8')
+    else:
+        raise NameError
```

Comment:
What happens if the resource identified by URI is a file, but is not readable by the user running the webserver?  What if it exists when we check if it's a directory, but then is deleted before we can list it?